031-set-pipeline-step/proposal.md

```diff
@@ -0,0 +1,120 @@
+
+When configuring an existing pipeline, however, the pipeline's paused status will not be changed. In other words, the `set_pipeline` step will leave already-existing paused pipelines in the paused state. The assumption here is that the pipeline has been manually paused by a pipeline operator, possibly in response to an emergent situation, and it should be left alone.
+
+## `((vars))` support
+
+Additionally, we should support `vars` (as in `fly set-pipeline -y`) and `var_files` (i.e. `fly set-pipeline -l`):
+
+```yaml
+plan:
+- get: ci
+- set_pipeline: release
+  file: ci/pipelines/release.yml
+  vars: {release_version: 5.3}
+  var_files:
+  - ci/pipelines/vars/foo.yml
+```
+
+# Open Questions
+
+* Should we prevent pipelines configured via a `set_pipeline` step from being
+  updated manually through `fly set-pipeline`?
+
+  Or should we emit a warning?
+
+* Should we support glob expansion in `var_files`?
+
+  The `concourse-pipeline` resource supports this by just performing glob
+  expansion against its local filesystem. For the `set_pipeline` step, this is
+  a bit more challenging - there *is* no local filesystem. Would we have to
+  implement glob expansion in the Baggageclaim API or something? How easily
+  would this translate to other runtimes?
+
+* `set_pipeline: self`
+
+  Currently, the `foo` in `set_pipeline: foo` is the name of a pipeline to set.
+  A pipeline could technically update itself by configuring its own name in the
+  step, but pipeline configs aren't meant to contain their own name, as doing
+  so prevents the config from being re-used as a 'pipeline template'.
+
+  Are self-updating pipelines a feature that we want to explicitly support by
+  allowing the keyword `self` in place of the pipeline name?
+
+  The [Projects RFC][projects-rfc] outlines a more 'git-ops' style flow for
+  configuring pipelines, where instead of having pipelines self-update they are
+  all configured in one central place (the project's `plan:`).
+
+  Pros:
+
+  * Fairly straightforward semantics which seem to support a natural follow-up
+    question after learning about the `set_pipeline` step.
+  * Keyword use has precedent in `version: every`/`version: latest`/`inputs: all`.
+
+  Cons:
+
+  * Supporting both self-updating pipelines and projects could cause confusion
+    and fragmentation; it doesn't seem wise to have two competing approaches to
+    the same goal.
+  * Given that `self` isn't *critical* (it's easy to work around through
+    templating, i.e. `set_pipeline: ((name))`), is it worth the
+    risk/maintenance?
+
+
+# Answered Questions
+
+none
+
+
+# New Implications
+
+## Deprecating `concourse-pipeline` resource
+
+Deprecating the `concourse-pipeline` resource should be the primary goal.
```

Comment:
The feature is very cool but has only a fraction of the scope and capabilities of the `concourse-pipeline` resource.  The area where they differ the most is for example the scope, `set_pipeline` is limited to the same team, unlike the `concourse-pipeline` resource, which allows us to create a pipeline in multiple teams and even different Concourse instances. 

We were thinking of the following use case:

Single multitenant Concourse instance (potentially using the external worker setup to limit teams to specific workers for security and performance reasons). With RBAC which limits people's ability to the basic stuff trigger jobs, start/pause pipelines, see the build results, and no hijacking, pruning and etc.

The idea here is to have a secure Concourse instance, that can handle multiple teams. We plan not to have a DB backup and treat the instance as a disposable unit. All content is defined as code and an onboarding pipeline which handles stuff like team creation, pipeline upload and etc. In the end, we would like to a definition on what should be put in Concourse and a sort of GitOps model, which is based on PR approval and follows immediate onboarding (team created, pipelines uploaded and started).

I know this is a weird scenario and maybe it is not a good idea, but never the less will give you a hint on the stupid things people are planning to do :D .

Reply:
I should clarify that while it's definitely a goal to deprecate the `concourse-pipeline` resource, it's *not* a goal for the `set_pipeline` step itself to be the sole thing to replace it and all of its functionality. It's just a piece of the puzzle, and I think it's important for that piece to stay small in scope to make it easier to recognize different patterns which may need to be addressed in different ways.

What you're planning to do sounds a lot like what I mentioned here: https://github.com/concourse/rfcs/pull/31#issuecomment-625273371 - I don't think it's weird at all, but I think it probably needs to be explored more in the context of https://github.com/concourse/rfcs/pull/32 which is a bit closer to the GitOps motivations.

Reply:
@vito spot on as always. Thanks for both hints. I'll try to figure out how this direction can potentially influence/impact our plans. In the past, we have been building tons of scripts to automate certain aspects of setting up individual teams environment and I would like to transition all that into a well-defined workflow.

Reply:
I wonder if we can unify Teams and Projects. 🤔 If we did this, you could theoretically have a full GitOps stack by configuring the "main project" in `concourse web` and having that project configure further projects with a `set_project` step, which in turn configure their own pipelines and access control (...and perhaps other projects?).

With all of this being driven by a build plan, you could do all sorts of interesting things like provisioning environments or worker pools.

I'll keep thinking about this and amend the Projects RFC. Thanks for your feedback!

Reply:
I think it's problematic to totally unify. Some folks prefer to think about attaching permissions to people (or roles, teams); others to assets (projects, folders, etc).

Under the hood it's fine to have some generic ACL representation, but I suspect most places think one way or the other.

Reply:
@jchesterpivotal I'm sure the devil is in the details - I'd like to figure out what those details are.

Those 'devilish details' will likely be around things that latch on to Teams for other functionality, like scoping credential access. But at least for credentials we already have a path forward with `var_sources`. Are there any other examples? 🤔

If you take away the name, a Team is really just a way of associating permissions to pipelines. I don't really see it as attaching permissions to people. (Unless I'm misunderstanding or forgetting something?)

Teams and Projects are both mechanisms for namespacing pipelines - to unify them, you could either add a permissions model to Projects or add the whole Projects functionality to Teams... which is to say that Teams on its own doesn't have much meat on its bones.

We constantly hear about folks trying to automate team/full-stack configuration:

* https://github.com/concourse/rfcs/pull/32#pullrequestreview-268542134
* https://github.com/concourse/rfcs/pull/31#issuecomment-625273371
* https://github.com/concourse/rfcs/pull/31#pullrequestreview-423266671 (this thread)
* https://github.com/concourse/concourse/issues/3208#issuecomment-603395182
  * Just about every time we hear people ask for API access or a browser-less logins flow, it's from this use case.

It feels like there's a lot of pressure pushing these concepts together, and it seems wise to try to limit the number of concepts with overlapping behavior anyway.

The Teams mechanic has evolved over time and never really had a strong identity. You used to configure all of a team's auth in `fly set-team` (including things like client IDs/secrets) and "log in to a team," but now you configure auth cluster-wide and log in as a user and just have access to whichever teams allow your user/group. This would pretty naturally extend to Projects, especially now that user tokens don't encode a list of teams (as of v6.1.0).

Folks are wildly inconsistent in how they use Teams - sometimes they're used to enforce boundaries, but sometimes they're literally just used to group pipelines together, with many teams given the same set of permissions. These different usage patterns and overall weak conceptual identity make Teams a difficult feature to design around; for example, the `fly` UX trips people up all the time because it assumes you'll be largely sticking to one team.

So I see this as an opportunity to rally around a replacement that has a lot of user demand and a more intentional design that lends itself well to Concourse-y GitOps patterns.

I'm jotting down notes and seeing a lot of things on the "pro" side, so forgive me if I'm talking past y'all and steamrolling a bit while this idea is still brewing - I'll try to find some time next week to amend the Projects RFC (or maybe fork it) since this discussion should probably move elsewhere. Curious to hear any thoughts on this until then either way. 🙂

Reply:
@vito when you say:

> having that project configure further projects with a set_project step, which in turn configure their own pipelines and access control

could you clarify what you mean by "configure their own access control"? Do you envision access control being defined by the project itself using some configuration file(s)?

If so, what do you think about the `set_project` step defining the access control for the project (much like how `fly set-team` currently defines the access control for the team). So, the main project + main project's access control is configured in `concourse web`, and could have a build plan like:

```
plan:
- set_project: A
  ...
  acl:
  - local_user: some_user
    role: owner
  - github:
      org: concourse
    role: viewer
```

and perhaps it inherits the parent's ACL as owner's of the project?

(full disclosure I don't fully know how RBAC is configured so this may not make sense)

This way, projects themselves don't need to change in scope (besides being able to run `set_project` steps) - much like the current proposal where projects exist within a team, they would just exist and have certain access controls associated to them.

Reply:
p.s. FWIW as someone who doesn't actually manage a Concourse deployment, let alone a multi-tenant one, I think this is a really cool idea!

Reply:
@aoldershaw Yeah - I can see either adding it to the `set_pipeline` step or adding the auth config to the project config itself.

Putting it in the project config would be more in line with how teams are structured now, where teams can update their auth config without requiring an operator.

I think the trade-off is that you would then commit your auth config into your own project repo - to some that's probably a feature, but there may be some situations where that's not ideal. (For example, you want the repo to be public but don't want to expose the identities of those who work on it.)

Reply:
I think @vito is onto something here. There's Teams as a way of bundling pipelines and then there's Teams as the sole interface for assigning roles / permissions to users. IMO the latter has always felt clunky and a bit bottom up; but we had to implement it this way when we rolled out Users & RBAC to minimize breaking changes. 

I think the way a user experiences Concourse through the web UI makes a lot more sense: you click login, you declare who you are, and Concourse tells you you can see/do. As an administrator, setting permissions is a bit backwards. You need to first identify the bundle of pipelines that you want to operate on, create that bundle through teams, create your permission matrix and then apply that through `set-teams`. For small-to-medium sized teams I think the permissioning is mostly binary: either a user can see/do everything on a team or it can't. For more mature teams with stricter rules, I think its more about exception setting, and that's almost always user-centric e.g. okay only Jane has the ability to `fly intercept`, everyone else is read only.

All of that is to say that permissions and access control on the team/project level probably won't make too much sense in the long term anyways. Whether it makes sense to combine team/projects for all the other reasons of project/team/pipeline automation though, i"ll leave that to folks with other opinons

Reply:
Whatever comes of it, I would be wary of reinventing any wheels, _especially_ in the permissions space. I prefer ACLs to RBACs due to their flexibility and precision, but RBACs have the nice property that they are easy to learn and reason about (they just suck for fine-grained permissions schemes).

The domain concepts here seem to be: user, thing that can be in a pipeline or is a pipeline, grouping of pipelines, operation on thing, permission linking subject-object-operation. The key being that a team/project unification makes a lot of sense as "grouping", and could even set itself up for arbitrary nesting, but that some additional effort for the permissions model would be needed to talk about operations and store the ACL triples.

Reply:
> With all of this being driven by a build plan, you could do all sorts of interesting things like provisioning environments or worker pools.

This one will be particularly of interest. We are moving into the direction of having dedicated "external" workers for each team as there is no other way right now to prevent potential security implications or even the noisy neighbor's considerations.

Reply:
As a follow-up, I talked with @tlwr in https://github.com/concourse/concourse/pull/5729 and we agreed that we can add an experimental `team:` field to expedite the deprecation of the `concourse-pipeline` resource while allowing us an way out so that it can be removed in the future. I've added this under the 'Experiments' section, along with `set_pipeline: self`.

@radoslav-tomov One additional thing to talk about there is also automatically *un*provisioning those environments when a team goes away. Probably a topic for another RFC though. 🙂